pkg/controller/csr/certificatesigningrequest.go

```diff
@@ -0,0 +1,225 @@
+package csr
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-version"
+	"github.com/libopenstorage/operator/drivers/storage/portworx/util"
+	"github.com/libopenstorage/operator/pkg/util/k8s"
+	coreops "github.com/portworx/sched-ops/k8s/core"
+	"github.com/sirupsen/logrus"
+	certv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ControllerName    = "px-csr-controller"
+	labelNameKey      = "name"
+	labelNameVal      = "portworx"
+	labelClusterIDKey = "cid"
+)
+
+var _ reconcile.Reconciler = &Controller{}
+
+var (
+	reMatchUserAnyNamespace   = regexp.MustCompile("system:serviceaccount:[^:]+:portworx")
+	minSupportedK8sVersion, _ = version.NewVersion("1.19.0")
+	registration              sync.Map
+)
+
+type Controller struct {
+	client   client.Client
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+	ctrl     controller.Controller
+}
+
+// Init initializes CSR controller
+func (c *Controller) Init(mgr manager.Manager) error {
+	if ver, err := k8s.GetVersion(); err != nil {
+		return fmt.Errorf("could not determine K8s version: %s", err)
+	} else if ver.LessThanOrEqual(minSupportedK8sVersion) {
+		logrus.Warnf("CSR auto-approve not supported on K8s version %s", ver)
+		c.ctrl = nil
+		c.client = nil
+		return nil
+	}
+
+	c.client = mgr.GetClient()
+	c.scheme = mgr.GetScheme()
+	c.recorder = mgr.GetEventRecorderFor(ControllerName)
+
+	var err error
+	// Create a new controller
+	c.ctrl, err = controller.New(ControllerName, mgr, controller.Options{Reconciler: c})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// StartWatch starts the watch on CertificateSigningRequests
+func (c *Controller) StartWatch() error {
+	if c.ctrl == nil || c.client == nil {
+		logrus.Warn("CSR.StartWatch() bailing out  (not initialized)")
+		return nil
+	}
+
+	err := c.ctrl.Watch(
+		&source.Kind{Type: &certv1.CertificateSigningRequest{}},
+		&handler.EnqueueRequestForObject{},
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile will automatically approve Portworx-requested certificates
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	var (
+		csr certv1.CertificateSigningRequest
+		ret reconcile.Result
+		log = logrus.WithField("Request.Name", req.Name)
+	)
+
+	if !util.IsTLSEnabled() {
+		log.Debugf("CSR.Reconcile() bailing out  (TLS not enabled)")
+		return ret, nil
+	}
+
+	log.Infof("Reconciling CSR")
```

Comment:
maybe not good to stomp this

Reply:
Uh.. I'm not sure what you mean by this..   Could you please elaborate?

Reply:
do we want to print this in every reconcile loop every 20s? don't know if that's necessary.

Reply:
Oh, I thought controller's `Reconcile()` triggers on K8s object's _changes_  (didn't realize it triggers on schedule, every 20 seconds).   If this is the case, what's the advantage of using K8s controllers in the fist place  (seems like cronjobs would do a similar job, and object watches would be more efficient?)

Other than this..
* the StorageCluster's `Reconcile()` will also drop an INFO-log every time it triggers
* also, since CSR-controller is a brand new functionality, I think it would be OK to have more verbose logging until we have better confidence the new code works w/o issues...